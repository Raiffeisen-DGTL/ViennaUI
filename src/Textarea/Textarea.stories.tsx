import React from 'react';
import { Story, Meta } from 'storybook';
import { Textarea, TextareaProps } from './Textarea';

export default {
    title: 'Development/Textarea',
    component: Textarea,
} as Meta;

export const Overview: Story<TextareaProps> = (args) => {
    const [value, setValue] = React.useState('');
    const handleChange = (e, data) => setValue(data.value);
    return <Textarea {...args} showCounter maxLength={150} value={value} onChange={handleChange} />;
};

export const WithAdaptive: Story<TextareaProps> = (args) => {
    const [value, setValue] = React.useState('text');
    const handleChange = (e, data) => setValue(data.value);
    return <Textarea {...args} size={{ base: 'xs', s: 's', m: 'xl' }} value={value} onChange={handleChange} />;
};
