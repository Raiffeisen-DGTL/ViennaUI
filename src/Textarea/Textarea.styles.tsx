import styled, { css } from 'styled-components';
import { getPresets } from '../Utils/styling';
import { Breakpoints, ResponsiveProp, responsivePreset } from '../Utils/responsiveness';

const presets = getPresets('textarea', {
    base: null,
    size: responsivePreset('$size', 'm'),
    design: '$design',
    hover: '$design',
    disabled: '$design',
    focused: '$design',
    invalid: '$design',
});

const placeholder = getPresets('textarea.placeholder', {
    base: null,
    disabled: null,
});

const counter = getPresets('textarea.counter', {
    base: null,
});

export const Root = styled.div`
    position: relative;
`;

export interface PropsBox<B = Breakpoints> {
    $design: 'outline' | 'material';
    $size: ResponsiveProp<'xs' | 's' | 'm' | 'l' | 'xl' | B>;
    $invalid?: boolean;
    $maxHeight?: number;
}
export const Box = styled.textarea<PropsBox>`
    vertical-align: middle;
    text-align: left;
    margin: 0;
    padding: 0;
    width: 100%;
    height: 100%;
    border: 0;
    flex: 1 1 auto;
    appearance: none;
    resize: none;

    ${presets.base}
    ${presets.size}
    ${presets.design}

    &:hover {
        ${presets.hover}
    }

    &:focus {
        ${presets.focused}
    }

    &:disabled,
    &:disabled::placeholder {
        ${presets.disabled}
    }

    ${({ $maxHeight }) =>
        $maxHeight &&
        css`
            max-height: ${$maxHeight}px;
        `}

    ${({ $invalid }) =>
        $invalid &&
        css`
            ${presets.invalid}
        `}


    &::-ms-clear,
    &::-ms-reveal {
        display: none;
    }

    &::-webkit-input-placeholder {
        ${placeholder.base}
    }

    &::-moz-placeholder {
        ${placeholder.base}
    }

    &:-ms-input-placeholder {
        ${placeholder.base}
    }
`;

export const Counter = styled.div`
    position: absolute;
    bottom: 0;
    right: 0;
    text-decoration: none;
    box-sizing: border-box;
    display: block;
    ${counter.base};
`;
