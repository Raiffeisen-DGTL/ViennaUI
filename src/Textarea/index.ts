export { Textarea, type TextareaProps, type TextareaEvent } from './Textarea';
