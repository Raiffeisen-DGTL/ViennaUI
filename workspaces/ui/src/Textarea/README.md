# Textarea

Компонент "Многострочное поле для ввода текста".
Используйте данный компонент в формах, когда пользователю необходимо ввести длинный (многострочный) текст.

## Импорт

```
import { Textarea } from 'vienna-ui';
```

## Свойства / Props

Свойства наследуются от [HTMLAttributes<HTMLTextAreaElement>](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L1746)

Prop | Type | Default | Description
--- | --- | --- | ---
value | string \| undefined | false |
size | "xs" \| "s" \| "m" \| "l" \| "xl" \| undefined | "m" |
design | "outline" \| "material" \| undefined | "outline" |
maxRows | number \| undefined | false | Максимальное число видимых строк
resize | boolean \| undefined | true | Возможность изменять вытосу компонента
invalid | boolean \| undefined | false |
onChange | TextareaEvent> \| undefined | false |
onClick | TextareaEvent> \| undefined | false |
onFocus | TextareaEvent> \| undefined | false |
onBlur | TextareaEvent> \| undefined | false |
onKeyDown | ((event: KeyboardEvent) => void) \| undefined | false |
onKeyPress | ((event: KeyboardEvent) => void) \| undefined | false |
onKeyUp | ((event: KeyboardEvent) => void) \| undefined | false |
ref | ((instance: HTMLTextAreaElement \| null) => void) \| RefObject \| null \| undefined | false |
autoCapitalize | string \| undefined | false | Автоматическая установка заглавной буквы
spellcheck | string \| undefined | false | Проверка орфографии

## Использование

По умолчанию дизайн `outline` и размер `m`

```
<Textarea placeholder='Текстовое поле' />
```

## Дизайн
##### Свойство `design`

Есть два доступных значения для свойства `design`: `outline` (по умолчанию) и `material`.

```jsx
<Textarea design='material' />
<Textarea design='outline' />
```

## Невалидное состояние
##### Свойство `invalid`

Невалидное поле

```
<Textarea placeholder='Текстовое поле' invalid />
```

## Неактивное состояние
##### Свойство `disabled`

Заблокированное поле

```
<Textarea value='Текстовое поле' disabled />
```

## Изменение высоты компонента
##### Свойство `resize`

Определяет возможность изменения высоты поля. По умолчанию `true`

```<Textarea placeholder='Текстовое поле' resize={false} />```

## Максимальное число видимых строк
##### Свойство `maxRows`

Максимальное значение количества видимых строк. По умолчанию равно `3`

```<Textarea placeholder='Текстовое поле' maxRows={1} />```
