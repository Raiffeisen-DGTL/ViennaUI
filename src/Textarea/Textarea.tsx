import React, { useRef, useEffect, useCallback, useState } from 'react';
import { Breakpoints } from '../Utils/responsiveness';
import { Root, Box, PropsBox, Counter } from './Textarea.styles';
import { Span } from '../Typography';
import { BoxStyled } from '../Utils/styled';
import { OnChangeHandler, Pretty } from '../Utils';
import { ViewOnly, WithViewOnly } from '@/ViewOnly';

export const defaultTextareaTestId: TextareaProps['testId'] = {
    counter: 'textarea_counter',
    maxCounter: 'textarea_max-counter',
};

export type TextareaEvent<T> = (event: T, data: { name?: string; value?: string }) => void;
export type TextareaPropsWithRef<B = Breakpoints> = TextareaProps<B> & React.RefAttributes<HTMLTextAreaElement>;

export interface TextareaProps<B = Breakpoints>
    extends Omit<BoxStyled<typeof Box, PropsBox>, 'onChange' | 'onClick' | 'onFocus' | 'onBlur' | 'ref'>,
        WithViewOnly {
    design?: PropsBox<B>['$design'];
    size?: PropsBox['$size'];
    invalid?: PropsBox<B>['$invalid'];
    value?: string;
    maxRows?: number;
    resize?: boolean;
    maxCounter?: number;
    showCounter?: boolean;
    testId?: {
        counter?: string;
        maxCounter?: string;
    };
    onChange?: OnChangeHandler<string, React.ChangeEvent<HTMLTextAreaElement>>;
    onClick?: TextareaEvent<React.MouseEvent<HTMLTextAreaElement>>;
    onFocus?: TextareaEvent<React.FocusEvent<HTMLTextAreaElement>>;
    onBlur?: TextareaEvent<React.FocusEvent<HTMLTextAreaElement>>;
    onKeyDown?: React.KeyboardEventHandler<HTMLTextAreaElement>;
    onKeyPress?: React.KeyboardEventHandler<HTMLTextAreaElement>;
    onKeyUp?: React.KeyboardEventHandler<HTMLTextAreaElement>;
}

const TextareaInternal = <B,>(props: TextareaProps<B>, ref: React.Ref<HTMLTextAreaElement>) => {
    const {
        disabled,
        invalid,
        size = 'm',
        rows = 3,
        cols = 50,
        design = 'outline',
        maxRows,
        onChange,
        onFocus,
        onBlur,
        resize = true,
        value,
        name,
        maxCounter,
        showCounter = false,
        maxLength,
        spellCheck = 'false',
        autoComplete = 'off',
        autoCorrect = 'off',
        wrap = 'soft',
        viewOnly,
        viewOnlyText,
        testId = defaultTextareaTestId,
        ...attrs
    } = props;
    const textareaEl = useRef<HTMLTextAreaElement | null>(null);
    const [charsCounter, setCharsCounter] = useState(value?.length ?? 0);
    const [maxHeight, setMaxHeight] = useState<undefined | number>(undefined);
    let cacheCalcStyles: {
        lineHeight: number;
        paddingTop: number;
        paddingBottom: number;
    };
    const invalidCharsCounter = maxCounter ? charsCounter > maxCounter : false;

    const counterColor = invalidCharsCounter ? 'moscow100' : 'seattle60';
    const doResize = useCallback(() => {
        // input is not defined or resizing is disabled
        if (!textareaEl || !textareaEl.current || !resize) {
            return;
        }

        // returns to default state
        textareaEl.current.style.height = 'auto';
        const border = textareaEl.current.offsetHeight - textareaEl.current.clientHeight;
        textareaEl.current.style.height = `${textareaEl.current.scrollHeight + border}px`;
    }, [resize, textareaEl, value]);

    useEffect(() => {
        doResize();
    }, [doResize]);

    useEffect(() => {
        if (value?.length) {
            setCharsCounter(value?.length);
        } else {
            setCharsCounter(0);
        }
    }, [value?.length]);

    const handleChangeEvent = useCallback(
        (event: React.ChangeEvent<HTMLTextAreaElement>) => {
            if (disabled) {
                return;
            }
            setCharsCounter(event.currentTarget.value?.length);
            if (onChange) {
                onChange({ value: event.currentTarget.value, event, options: { name } });
            }
            doResize();
        },
        [disabled, onChange]
    );

    const handleFocus = useCallback(
        (event: React.FocusEvent<HTMLTextAreaElement>) => {
            if (onFocus) {
                onFocus(event, { value: event.currentTarget.value, name });
            }
        },
        [onFocus]
    );

    const handleBlur = useCallback(
        (event: React.FocusEvent<HTMLTextAreaElement>) => {
            if (onBlur) {
                onBlur(event, { value: event.currentTarget.value, name });
            }
        },
        [onBlur]
    );

    const getTextareaStyles = (
        node: HTMLTextAreaElement
    ): {
        lineHeight: number;
        paddingTop: number;
        paddingBottom: number;
    } => {
        if (cacheCalcStyles) {
            return cacheCalcStyles;
        }

        const computedStyles = window.getComputedStyle(node);

        cacheCalcStyles = {
            lineHeight: parseInt(computedStyles.getPropertyValue('line-height'), 10),
            paddingTop: parseInt(computedStyles.getPropertyValue('padding-top'), 10),
            paddingBottom: parseInt(computedStyles.getPropertyValue('padding-bottom'), 10),
        };

        return cacheCalcStyles;
    };

    const setMaxRows = useCallback(() => {
        // input is not defined or resizing is disabled
        if (!textareaEl || !textareaEl.current || !maxRows) {
            return;
        }

        const { lineHeight, paddingTop, paddingBottom } = getTextareaStyles(textareaEl.current);
        setMaxHeight(lineHeight * maxRows + paddingTop + paddingBottom);
    }, [maxRows]);

    useEffect(() => {
        setMaxRows();
    }, [setMaxRows]);

    useEffect(() => {
        if (typeof ref === 'function') {
            ref(textareaEl.current);
        } else if (ref && 'current' in ref) {
            (ref as React.MutableRefObject<HTMLTextAreaElement | null>).current = textareaEl.current;
        }
    });
    const sizesInCounter = size === 'l' ? 'm' : 's';

    if (viewOnly) {
        return <ViewOnly size={size}>{viewOnlyText ?? value}</ViewOnly>;
    }

    return (
        <Root>
            <Box
                {...attrs}
                $maxHeight={maxHeight}
                $invalid={invalid ?? invalidCharsCounter}
                $design={design}
                $size={size}
                $showCounter={showCounter}
                value={value}
                disabled={disabled}
                aria-invalid={!!invalid || invalidCharsCounter}
                rows={rows}
                cols={cols}
                ref={textareaEl}
                name={name}
                spellCheck={spellCheck}
                autoComplete={autoComplete}
                autoCorrect={autoCorrect}
                wrap={wrap}
                maxLength={maxLength}
                onChange={handleChangeEvent}
                onFocus={handleFocus}
                onBlur={handleBlur}
            />
            {showCounter && (
                <Counter>
                    <Span size={sizesInCounter} color={counterColor} data-testid={testId?.counter}>
                        {charsCounter}
                    </Span>
                    {maxCounter && (
                        <Span size={sizesInCounter} color={counterColor} data-testid={testId?.maxCounter}>
                            /{maxCounter}
                        </Span>
                    )}
                </Counter>
            )}
        </Root>
    );
};

export namespace Textarea {
    export type OnChange = Pretty.Func<OnChangeHandler<string, React.ChangeEvent<HTMLTextAreaElement>>>;
    export type OnClick = Pretty.Func<TextareaEvent<React.MouseEvent<HTMLTextAreaElement>>>;
    export type OnFocus = Pretty.Func<TextareaEvent<React.FocusEvent<HTMLTextAreaElement>>>;
    export type OnBlur = Pretty.Func<TextareaEvent<React.FocusEvent<HTMLTextAreaElement>>>;
}

export const Textarea = React.forwardRef(TextareaInternal) as (<B>(
    props: TextareaPropsWithRef<B>
) => ReturnType<typeof TextareaInternal>) & {
    displayName?: string;
};

Textarea.displayName = 'Textarea';
