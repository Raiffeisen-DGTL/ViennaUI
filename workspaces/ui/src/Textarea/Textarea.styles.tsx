import styled, { css } from 'styled-components';
import { getPresets } from 'vienna.ui-primitives';
import { Design, Size } from './Textarea';

const presets = getPresets('textarea', {
    base: null,
    size: 'size',
    design: 'design',
    hover: 'design',
    disabled: 'design',
    focused: 'design',
    invalid: 'design',
    placeholder: 'design',
});

interface BoxProps {
    readonly design: Design;
    readonly size: Size;
    readonly invalid?: boolean;
    readonly maxHeight?: number;
}

export const Box = styled.textarea<BoxProps>`
    vertical-align: middle;
    text-align: left;
    margin: 0;
    padding: 0;
    width: 100%;
    height: 100%;
    border: 0;
    flex: 1 1 auto;
    appearance: none;
    resize: none;

    ${presets.base}
    ${presets.size}
    ${presets.design}

    &:hover {
        ${presets.hover}
    }

    &:focus {
        ${presets.focused}
    }

    &:disabled,
    &:disabled::placeholder {
        ${presets.disabled}
    }

    ${(props) =>
        props.maxHeight &&
        css`
            max-height: ${props.maxHeight}px;
        `}

    ${(props) =>
        props.invalid &&
        css`
            ${presets.invalid}
        `}

    &::-ms-clear,
    &::-ms-reveal {
        display: none;
    }

    &:focus::-webkit-input-placeholder {
        color: transparent;
    }

    &:focus::-moz-placeholder {
        color: transparent;
    }

    &:focus:-ms-input-placeholder {
        color: transparent;
    }
`;
