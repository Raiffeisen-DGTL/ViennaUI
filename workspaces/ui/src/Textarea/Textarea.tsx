import React, { useRef, useEffect, useCallback, useState, ForwardRefExoticComponent } from 'react';
import { Box } from './Textarea.styles';

export type TextareaEvent<T> = (event: T, data?: { name?: string; value?: string }) => void;
export type Size = 'xs' | 's' | 'm' | 'l' | 'xl';
export type Design = 'outline' | 'material';

interface Props {
    value?: string;
    size?: Size;
    design?: Design;
    rows?: number;
    maxRows?: number;
    resize?: boolean;
    disabled?: boolean;
    invalid?: boolean;
    onChange?: TextareaEvent<React.FormEvent<HTMLTextAreaElement>>;
    onClick?: TextareaEvent<React.FormEvent<HTMLTextAreaElement>>;
    onFocus?: TextareaEvent<React.FormEvent<HTMLTextAreaElement>>;
    onBlur?: TextareaEvent<React.FormEvent<HTMLTextAreaElement>>;
    onKeyDown?: React.KeyboardEventHandler<HTMLTextAreaElement>;
    onKeyPress?: React.KeyboardEventHandler<HTMLTextAreaElement>;
    onKeyUp?: React.KeyboardEventHandler<HTMLTextAreaElement>;
    ref?: React.Ref<HTMLTextAreaElement>;
}

interface HTMLAttributeProps {
    id?: string;
    name?: string;
    cols?: number;
    wrap?: 'hard' | 'soft' | 'off';
    autoComplete?: string;
    autoCapitalize?: string;
    autoCorrect?: string;
    autoFocus?: boolean;
    maxLength?: number;
    minLength?: number;
    placeholder?: string;
    readOnly?: boolean;
    required?: boolean;
    tabIndex?: number;
    spellcheck?: string;
}

export type TextareaProps = Props & HTMLAttributeProps;

export const Textarea: React.FC<TextareaProps> = React.forwardRef(
    (props: TextareaProps, ref: React.Ref<HTMLTextAreaElement>) => {
        const {
            disabled,
            invalid,
            size,
            rows,
            design = 'outline',
            maxRows,
            onChange,
            onFocus,
            onBlur,
            resize,
            value,
            name,
            ...attrs
        } = props;
        const textareaEl = useRef<HTMLTextAreaElement>(null);
        const [maxHeight, setMaxHeight] = useState<undefined | number>(undefined);
        let cacheCalcStyles: any = false;

        const doResize = useCallback(() => {
            // input is not defined or resizing is disabled
            if (!textareaEl || !textareaEl.current || !resize) {
                return;
            }

            // returns to default state
            textareaEl.current.style.height = 'auto';

            const border = textareaEl.current.offsetHeight - textareaEl.current.clientHeight;
            textareaEl.current.style.height = `${textareaEl.current.scrollHeight + border}px`;
        }, [resize, textareaEl, value]);

        useEffect(() => {
            doResize();
        }, [doResize]);

        const handleChangeEvent = useCallback(
            (event: React.FormEvent<HTMLTextAreaElement>) => {
                if (disabled) {
                    return;
                }

                if (onChange) {
                    onChange(event, { value: event.currentTarget.value, name });
                }

                doResize();
            },
            [disabled, onChange]
        );

        const handleFocus = useCallback(
            (event: React.FormEvent<HTMLTextAreaElement>) => {
                if (onFocus) {
                    onFocus(event, { value: event.currentTarget.value, name });
                }
            },
            [onFocus]
        );

        const handleBlur = useCallback(
            (event: React.FormEvent<HTMLTextAreaElement>) => {
                if (onBlur) {
                    onBlur(event, { value: event.currentTarget.value, name });
                }
            },
            [onBlur]
        );

        const getTextareaStyles = (
            node: HTMLTextAreaElement
        ): {
            lineHeight: number;
            paddingTop: number;
            paddingBottom: number;
        } => {
            if (cacheCalcStyles) {
                return cacheCalcStyles;
            }

            const computedStyles = window.getComputedStyle(node);

            cacheCalcStyles = {
                lineHeight: parseInt(computedStyles.getPropertyValue('line-height'), 10),
                paddingTop: parseInt(computedStyles.getPropertyValue('padding-top'), 10),
                paddingBottom: parseInt(computedStyles.getPropertyValue('padding-bottom'), 10),
            };

            return cacheCalcStyles;
        };

        const setMaxRows = useCallback(() => {
            // input is not defined or resizing is disabled
            if (!textareaEl || !textareaEl.current || !maxRows) {
                return;
            }

            const { lineHeight, paddingTop, paddingBottom } = getTextareaStyles(textareaEl.current);
            setMaxHeight(lineHeight * maxRows + paddingTop + paddingBottom);
        }, [maxRows]);

        useEffect(() => {
            setMaxRows();
        }, [setMaxRows]);

        useEffect(() => {
            if (typeof ref === 'function') {
                ref(textareaEl.current);
            } else if (ref && 'current' in ref) {
                (ref as any).current = textareaEl.current;
            }
        });

        return (
            <Box
                {...attrs}
                maxHeight={maxHeight}
                value={value}
                disabled={disabled}
                invalid={invalid}
                aria-invalid={!!invalid}
                design={design}
                size={size as Size}
                rows={rows}
                ref={textareaEl}
                name={name}
                onChange={handleChangeEvent}
                onFocus={handleFocus}
                onBlur={handleBlur}
            />
        );
    }
) as ForwardRefExoticComponent<TextareaProps>;

Textarea.displayName = 'Textarea';
Textarea.defaultProps = {
    size: 'm',
    rows: 3,
    cols: 50,
    resize: true,
    spellcheck: 'false',
    autoComplete: 'off',
    autoCorrect: 'off',
    wrap: 'soft',
    design: 'outline',
};
export default Textarea;
