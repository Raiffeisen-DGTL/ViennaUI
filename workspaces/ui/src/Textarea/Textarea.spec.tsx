import React from 'react';
import Textarea, { TextareaProps } from './Textarea';

test('Textarea', async () => {
    const component = <Textarea />;

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();
});

test('Textarea w/ size', async () => {
    const sizes: TextareaProps['size'][] = ['xs', 's', 'm', 'l', 'xl'];
    const components: JSX.Element[] = [];

    sizes.forEach((size) => components.push(<Textarea key={size} size={size} />));

    const snap = snapshot.render(components);
    expect(snap).toMatchSnapshot();
});

test('Textarea w/ design', async () => {
    const designs: TextareaProps['design'][] = ['outline', 'material'];
    const components: JSX.Element[] = [];

    designs.forEach((design) => components.push(<Textarea key={design} design={design} />));

    const snap = snapshot.render(components);
    expect(snap).toMatchSnapshot();
});

test('Textarea w/ value', async () => {
    const component = <Textarea value='some value' />;

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();
});

test('Textarea w/ disabled', async () => {
    const designs: TextareaProps['design'][] = ['outline', 'material'];
    const components: JSX.Element[] = [];

    designs.forEach((design) => components.push(<Textarea key={design} design={design} disabled />));

    const snap = snapshot.render(components);
    expect(snap).toMatchSnapshot();
});

test('Textarea w/ invalid', async () => {
    const designs: TextareaProps['design'][] = ['outline', 'material'];
    const components: JSX.Element[] = [];

    designs.forEach((design) => components.push(<Textarea key={design} design={design} invalid />));

    const snap = snapshot.render(components);
    expect(snap).toMatchSnapshot();
});

test('Textarea w/ onChange()', () => {
    const fn = jest.fn();
    const event = { currentTarget: { value: 'value' } };

    const snap = snapshot.shallow(<Textarea name='textarea' onChange={fn} />);

    snap.simulate('change', event);
    expect(snap).toMatchSnapshot();

    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledWith(event, { name: 'textarea', value: 'value' });
});

test('Textarea w/ onChange() w/ disabled', () => {
    const fn = jest.fn();
    const event = { currentTarget: { value: 'value' } };

    const snap = snapshot.shallow(<Textarea disabled name='textarea' onChange={fn} />);

    snap.simulate('change', event);
    expect(snap).toMatchSnapshot();

    expect(fn).toHaveBeenCalledTimes(0);
});

test('Textarea w/ onFocus()', () => {
    const fn = jest.fn();
    const event = { currentTarget: { value: 'value' } };

    const snap = snapshot.shallow(<Textarea name='textarea' onFocus={fn} />);

    snap.simulate('focus', event);
    expect(snap).toMatchSnapshot();

    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledWith(event, { name: 'textarea', value: 'value' });
});

test('Textarea w/ onBlur()', () => {
    const fn = jest.fn();
    const event = { currentTarget: { value: 'value' } };

    const snap = snapshot.shallow(<Textarea name='textarea' onBlur={fn} />);

    snap.simulate('blur', event);
    expect(snap).toMatchSnapshot();

    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledWith(event, { name: 'textarea', value: 'value' });
});

test('Textarea w/ onKeyDown()', () => {
    const fn = jest.fn();
    const event = { key: 'a' };

    const snap = snapshot.shallow(<Textarea name='textarea' onKeyDown={fn} />);

    snap.simulate('keydown', event);
    expect(snap).toMatchSnapshot();

    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledWith(event);
});

test('Textarea w/ onKeyPress()', () => {
    const fn = jest.fn();
    const event = { key: 'a' };

    const snap = snapshot.shallow(<Textarea name='textarea' onKeyPress={fn} />);

    snap.simulate('keypress', event);
    expect(snap).toMatchSnapshot();

    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledWith(event);
});

test('Textarea w/ onKeyUp()', () => {
    const fn = jest.fn();
    const event = { key: 'a' };

    const snap = snapshot.shallow(<Textarea name='textarea' onKeyUp={fn} />);

    snap.simulate('keyup', event);
    expect(snap).toMatchSnapshot();

    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledWith(event);
});

test('Textarea w/ some props', () => {
    const props = {
        className: 'className',
        id: 'id',
        title: 'title',
        onClick: () => 'Click',
    };

    const component = <Textarea {...props} name='textarea' />;

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();

    const wrapper = snapshot.shallow(component);
    for (const [key, value] of Object.entries(props)) {
        expect(wrapper.props()[key]).toEqual(value);
    }
});
