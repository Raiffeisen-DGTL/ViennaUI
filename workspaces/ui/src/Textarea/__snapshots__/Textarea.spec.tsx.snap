// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea 1`] = `
<textarea
  aria-invalid="false"
  autocomplete="off"
  autocorrect="off"
  class="Textareastyles__Box-hlpio7-0 caKrog"
  cols="50"
  rows="3"
  wrap="soft"
/>
`;

exports[`Textarea w/ design 1`] = `
Array [
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 caKrog"
    cols="50"
    rows="3"
    wrap="soft"
  />,
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 fYDlcN"
    cols="50"
    rows="3"
    wrap="soft"
  />,
]
`;

exports[`Textarea w/ disabled 1`] = `
Array [
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 caKrog"
    cols="50"
    disabled=""
    rows="3"
    wrap="soft"
  />,
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 fYDlcN"
    cols="50"
    disabled=""
    rows="3"
    wrap="soft"
  />,
]
`;

exports[`Textarea w/ invalid 1`] = `
Array [
  <textarea
    aria-invalid="true"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 hdLoNw"
    cols="50"
    rows="3"
    wrap="soft"
  />,
  <textarea
    aria-invalid="true"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 fBVaZQ"
    cols="50"
    rows="3"
    wrap="soft"
  />,
]
`;

exports[`Textarea w/ onBlur() 1`] = `
<Textareastyles__Box
  aria-invalid={false}
  autoComplete="off"
  autoCorrect="off"
  cols={50}
  design="outline"
  name="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  rows={3}
  size="m"
  spellcheck="false"
  wrap="soft"
/>
`;

exports[`Textarea w/ onChange() 1`] = `
<Textareastyles__Box
  aria-invalid={false}
  autoComplete="off"
  autoCorrect="off"
  cols={50}
  design="outline"
  name="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  rows={3}
  size="m"
  spellcheck="false"
  wrap="soft"
/>
`;

exports[`Textarea w/ onChange() w/ disabled 1`] = `
<Textareastyles__Box
  aria-invalid={false}
  autoComplete="off"
  autoCorrect="off"
  cols={50}
  design="outline"
  disabled={true}
  name="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  rows={3}
  size="m"
  spellcheck="false"
  wrap="soft"
/>
`;

exports[`Textarea w/ onFocus() 1`] = `
<Textareastyles__Box
  aria-invalid={false}
  autoComplete="off"
  autoCorrect="off"
  cols={50}
  design="outline"
  name="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  rows={3}
  size="m"
  spellcheck="false"
  wrap="soft"
/>
`;

exports[`Textarea w/ onKeyDown() 1`] = `
<Textareastyles__Box
  aria-invalid={false}
  autoComplete="off"
  autoCorrect="off"
  cols={50}
  design="outline"
  name="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyDown={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "key": "a",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  rows={3}
  size="m"
  spellcheck="false"
  wrap="soft"
/>
`;

exports[`Textarea w/ onKeyPress() 1`] = `
<Textareastyles__Box
  aria-invalid={false}
  autoComplete="off"
  autoCorrect="off"
  cols={50}
  design="outline"
  name="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyPress={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "key": "a",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  rows={3}
  size="m"
  spellcheck="false"
  wrap="soft"
/>
`;

exports[`Textarea w/ onKeyUp() 1`] = `
<Textareastyles__Box
  aria-invalid={false}
  autoComplete="off"
  autoCorrect="off"
  cols={50}
  design="outline"
  name="textarea"
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  onKeyUp={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "key": "a",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  rows={3}
  size="m"
  spellcheck="false"
  wrap="soft"
/>
`;

exports[`Textarea w/ size 1`] = `
Array [
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 KCehh"
    cols="50"
    rows="3"
    wrap="soft"
  />,
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 jJHmpZ"
    cols="50"
    rows="3"
    wrap="soft"
  />,
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 caKrog"
    cols="50"
    rows="3"
    wrap="soft"
  />,
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 caKrog"
    cols="50"
    rows="3"
    wrap="soft"
  />,
  <textarea
    aria-invalid="false"
    autocomplete="off"
    autocorrect="off"
    class="Textareastyles__Box-hlpio7-0 caKrog"
    cols="50"
    rows="3"
    wrap="soft"
  />,
]
`;

exports[`Textarea w/ some props 1`] = `
<textarea
  aria-invalid="false"
  autocomplete="off"
  autocorrect="off"
  class="Textareastyles__Box-hlpio7-0 caKrog className"
  cols="50"
  id="id"
  name="textarea"
  rows="3"
  title="title"
  wrap="soft"
/>
`;

exports[`Textarea w/ value 1`] = `
<textarea
  aria-invalid="false"
  autocomplete="off"
  autocorrect="off"
  class="Textareastyles__Box-hlpio7-0 caKrog"
  cols="50"
  rows="3"
  wrap="soft"
>
  some value
</textarea>
`;
